### ISSUE TITLE
Ceramic fails to start "interactive" on Windows / sbcl

### ISSUE THREAD

**Issue by dertuxmalwieder** [ceramic/ceramic#25]
While the CCL issues are being worked on, here's a new one:

When I use Ceramic inside SBCL on Windows (64-bit), `(ceramic:interactive)` tries to run `/usr/bin/env` after having downloaded Electron and subsequently, of course, fails:

```
(RUN-PROGRAM "/usr/bin/env" ("C:/Users/hp/.ceramic/electron/electron.exe") :ENV NIL :ENVIRONMENT NIL :WAIT NIL :SEARCH T :INPUT :STREAM :IF-INPUT-DOES-NOT-EXIST NIL :OUTPUT :STREAM :IF-OUTPUT-EXISTS :ERROR :ERROR :OUTPUT :IF-ERROR-EXISTS :ERROR :STATUS-HOOK NIL :EXTERNAL-FORMAT :DEFAULT :DIRECTORY NIL)
```

**Comment by eudoxia0** [comment]
Problem is here https://github.com/sellout/external-program/issues/20

**Comment by dertuxmalwieder** [comment]
Oh, sweet. Anything you can do about that?

**Comment by eudoxia0** [comment]
Apparently it was fixed but never merged:

https://github.com/sellout/external-program/pull/18

**Comment by eudoxia0** [comment]
It is merged

**Comment by dertuxmalwieder** [comment]
Great. Waiting for QL update now. Thanks. :)

**Comment by dertuxmalwieder** [comment]
I made a `git clone` of the current `external-program` code. It does not crash anymore, but it, uhm ... calls `electron.exe` and does nothing anymore?

![Screenshot](http://i.imgur.com/zbZWMX8.png)

The only input that reacts now is Ctrl-C.

**Comment by eudoxia0** [comment]
Huh. Can you call `external-program:start` with the path to the `electron.exe` and no arguments, and see if you can get the process object?

And then also try calling:

``` lisp
(ceramic.electron:start-process (ceramic.electron:release-directory) :operating-system ceramic.os:*operating-system*)
```

**Comment by dertuxmalwieder** [comment]
[**@eudoxia0**](https://github.com/eudoxia0) commented on [17. Nov. 2015, 00:43 MEZ](https://github.com/ceramic/ceramic/issues/25#issuecomment-157210810):

> Huh. Can you call `external-program:start` with the path to the `electron.exe` and no arguments, and see if you can get the process object?

It requires 2 parameters...?

> And then also try calling:
> 
> <div class="highlight highlight-source-lisp"><pre>(ceramic.electron:start-process (ceramic.electron:release-directory) :operating-system ceramic.os:<span class="pl-smi">*operating-system*</span>)</pre></div>

Sure. I get the same result: A waiting ("READY") subprocess.

**Comment by eudoxia0** [comment]
Update: nothing related to this issue just to say the build and tests are passing again.

**Comment by eudoxia0** [comment]
> It requires 2 parameters...?

Yeah, the second one is the list of arguments, that can be `NIL`.

**Comment by eudoxia0** [comment]
> Sure. I get the same result: A waiting ("READY") subprocess.

Okay, the relevant code is here:

``` lisp
(defun wait-for-startup (process)
  "Wait until the Electron process has started and is ready to go."
  (labels ((read-all-from-stream (stream)
             (concatenate 'string
                          (loop for byte = (read-char-no-hang stream nil nil)
                                while byte collecting byte)))
           (process-stdout ()
             (read-all-from-stream
              (external-program:process-output-stream
               process))))
    (ceramic.log:log-message "Waiting for startup...")
    (let ((output (process-stdout)))
      (loop until (search "READY" output) do
        (let ((new-output (process-stdout)))
          (setf output (concatenate 'string output new-output)))))
    ;; Clear all stdout
    (process-stdout)
    (ceramic.log:log-message "Electron process started.")
    t))
```

Try deleting that final call to `process-stdout` right before `log-message` and see if that changes anything.

**Comment by dertuxmalwieder** [comment]
It still is stuck with the same message for me. :(

**Comment by eudoxia0** [comment]
Worst case scenario this is something on the Node side and has something to do with Windows newlines.

I suppose there's one way to debug that: from a `cmd.exe`, run the Electron executable, and start sending commands manually. This is what you would want to send:

``` json
{"cmd":"create-window","name":"win","show":false,"title":"My Window","resizable":false}
```

Followed by:

``` json
{"cmd":"show-window","name":"win"}.
```

That _should_ work, if the Node code is portable.

**Comment by dertuxmalwieder** [comment]
Running Electron.exe immediately sends it to the background (?); I get "READY" and the command prompt. I have no clue how I can send anything to the process...

**Comment by eudoxia0** [comment]
Well, that's a problem. But I guess it makes sense that the Electron executable would run in the background.

**Comment by dertuxmalwieder** [comment]
That explains the "READY" though.

So the problem obviously is that Ceramic expects a  foreground Electron. :-)

**Comment by eudoxia0** [comment]
Yes. On Linux Electron is foreground. Let me test on my Windows VM.

**Comment by dertuxmalwieder** [comment]
Windows 10 here, grabbing Electron 0.28. Maybe updating Electron to a newer version (when downloading it) could be a solution?

**Comment by eudoxia0** [comment]
I'm looking at the command line options right now to see if there's anything.

**Comment by dertuxmalwieder** [comment]
I'm off to bed for now. Will test tomorrow!

**Comment by eudoxia0** [comment]
Good night, and thanks for your patience.

**Comment by dertuxmalwieder** [comment]
No problem. I know how hard it is to track down errors on platforms you rarely use.

**Comment by eudoxia0** [comment]
Well, latest Electron runs in the background on Windows, and I doubt there's anything I can do.

**Comment by dertuxmalwieder** [comment]
Should we file a bug there? Could you? Would it make sense?

**Comment by eudoxia0** [comment]
I think I'm going to ask around IRC or something first.

**Comment by eudoxia0** [comment]
Aw crap they have a web forum. Sigh. Guess I have to manager another password.

**Comment by eudoxia0** [comment]
Oh hey look at this https://discuss.atom.io/t/foreground-option-does-nothing/16478

**Comment by dertuxmalwieder** [comment]
Don't they just manage issues on Github? 

Hm, the link might not be related to Electron, Electron does not seem to have any options for that...?

**Comment by eudoxia0** [comment]
Yeah, but I thought I should ask a question on a forum before opening an issue.

**Comment by dertuxmalwieder** [comment]
Hmm... does Ceramic have a forum? 

Seriously, mostly those forums are crowded by a lot of people, and none of them are devs.

**Comment by eudoxia0** [comment]
No.

The issue tracker is a pseudo-forum.

**Comment by dertuxmalwieder** [comment]
I see. 

An issue tracker is for placing issues you have stumbled across. What can forums do? Discuss the issue before it's placed there eventually? :) 

Maybe I'm too cynical for that.

**Comment by eudoxia0** [comment]
Well, you can also ask questions. Sadly non-contributors can't tag their own issues with the `question` label.

**Comment by eudoxia0** [comment]
Aha! On Windows, you have to set `ELECTRON_ENABLE_LOGGING` to `YES` and then the program launches in the foreground.

**Comment by dertuxmalwieder** [comment]
And it probably logs things. (Where and which things?) 

Can Ceramic automatize that? 

(I'll check that tomorrow. Thanks for your efforts. Good night for now.)

**Comment by eudoxia0** [comment]
Automatize what, setting the environment variable, or extracting logging info? Probably and I guess, respectively.

Good night.

**Comment by dertuxmalwieder** [comment]
Hm, the logging info could be interesting for us too. Depends on what it logs. But I actually thought the user should not have to think about environment variables when installing Lisp systems.

**Comment by eudoxia0** [comment]
Sure, of course, Ceramic should set that on Windows.

**Comment by dertuxmalwieder** [comment]
I've set `ELECTRON_ENABLE_LOGGING` to `YES`, Ceramic still ends with "READY".

**Comment by eudoxia0** [comment]
Oh my godddddddddddddddddddddddddddddddddddddddddddd

**Comment by dertuxmalwieder** [comment]
Well, sorry.

**Comment by eudoxia0** [comment]
Please don't think I'm blaming you :) this is my fault for releasing things into the wild without even firing up a VM. I'm gonna copy the `main.js` to my Electron download on the Windows VM and see what I can find out.

**Comment by dertuxmalwieder** [comment]
Hey, it _is_ my fault. I try to do serious development on Windows.

Jokes aside, if I can help you, tell me.

**Comment by eudoxia0** [comment]
Well, I can reproduce it here. Time to dig deeper.

**Comment by dertuxmalwieder** [comment]
Heh. OK, tell me when you **can't** reproduce it anymore so I can test. ;-)

**Comment by eudoxia0** [comment]
Hmm, maybe if I try running it as a regular process.

**Comment by eudoxia0** [comment]
Nope that doesn't work.

**Comment by dertuxmalwieder** [comment]
:(

**Comment by eudoxia0** [comment]
Well, I caved and raised an issue.

**Comment by dertuxmalwieder** [comment]
Oh, the first reply is sad.

**Comment by anaisbetts** [comment]
@dertuxmalwieder It wasn't my idea, sorry! I'm just the messenger :)   Fortunately, writing this kind of wrapper in node.js is super super easy, just call `spawn`

**Comment by dertuxmalwieder** [comment]
How does Atom do it?

I prefer not to use Node.js...

**Comment by eudoxia0** [comment]
I suppose it might be possible to capture a process's IO streams even if it goes into the background, though I'm not sure how you'd do it.

**Comment by anaisbetts** [comment]
Yeah, so that part is the same, the only difference is whether the parent console will block the console until exit or not - STDIO works the same

**Comment by eudoxia0** [comment]
Maybe we could invert the architecture, and have the Node.js process launch the Common Lisp binary, instead of the converse.

**Comment by eudoxia0** [comment]
Alternatively: replace this awful stdio-based solution with a more stable HTTP server.

**Comment by dertuxmalwieder** [comment]
Do you have a plan yet?

**Comment by eudoxia0** [comment]
No.

**Comment by eudoxia0** [comment]
Update: the `remote-js` branch has a functioning WebSockets-based communication system.

**Comment by dertuxmalwieder** [comment]
Sweet. What should I test?

**Comment by eudoxia0** [comment]
If you can run the tests on that branch on Windows then it's working fine.

**Comment by Symbolics** [comment]
Is this working for anyone? I just got the latest code from the master branch and after quickloading "successfully" I get:

```
CL-USER(5): (ceramic:start)
Ceramic: Starting server...
Ceramic: Starting Electron process...

debugger invoked on a SIMPLE-ERROR in thread
#<THREAD "main thread" RUNNING {1002300083}>:
  Couldn't execute "C:\\Users\\symbolics\\.ceramic\\electron\\electron.exe": The system cannot find the file specified.

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [ABORT] Exit debugger, returning to top level.

(RUN-PROGRAM "C:/Users/symbolics/.ceramic/electron/electron.exe" ("C:/Users/symbolics/.ceramic/electron/resources/default_app/" "localhost" "40000") :ENV NIL :ENVIRONMENT NIL :WAIT NIL :SEARCH T :INPUT NIL :IF-INPUT-DOES-NOT-EXIST NIL :OUTPUT #<SYNONYM-STREAM :SYMBOL SB-SYS:*STDOUT* {100015E733}> :IF-OUTPUT-EXISTS :ERROR :ERROR :OUTPUT :IF-ERROR-EXISTS :ERROR :STATUS-HOOK NIL :EXTERNAL-FORMAT :DEFAULT :DIRECTORY NIL :ESCAPE-ARGUMENTS T :WINDOW NIL)
0]
```

and indeed the .ceramic directory does not exist, and of neither does electron.exe. This is on the SBCL 2.0.10 and the latest Windows 10, started from the command line. I've got electron installed and working for other purposes and it would be nice if ceramic could use the node.js installation: `C:/Program Files/nodejs`

Anyone got this to work on MS Windows recently? Hopefully the 0.2 milestone will fix it...
